### ISSUE TITLE
EqualConstraint does Not use Equals Override on the Expected Object

### ISSUE THREAD

**Issue by lundmikkel** [nunit/nunit#1439]
According to [the documentation](https://github.com/nunit/docs/wiki/EqualConstraint#user-content-notes), the `EqualConstraint` uses the compared objects' `Equals()` methods:

> 1. When checking the equality of user-defined classes, NUnit first examines each class to determine whether it implements IEquatable<T>. If either object implements the interface for the type of the other object, then that implementation is used in making the comparison. If neither class implements the appropriate interface, NUnit makes use of the Equals override on the expected object. [...]

I cannot get that behavior in my test. As an example, [I have created a minimal working example](https://github.com/lundmikkel/nunit_issue_1439) that demonstrates my problem. Here is a brief description off what I did and what I expect:

I have an interface:

```c#
public interface IInterface<T> : IEnumerable<T> {
    int Count { get; }
}
```
Its an enumerable that knows its length. I also have two different classes that implement the interface, `Class1` and `Class2`:
```c#
public class Class1<T> : IInterface<T> {
    private readonly IEnumerable<T> _enumerable;

    public Class1(IEnumerable<T> enumerable) {
        var array = enumerable as T[] ?? enumerable.ToArray();
        _enumerable = array;
        Count = array.Length;
    }

    public IEnumerator<T> GetEnumerator() => _enumerable.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public int Count { get; }
}

public class Class2<T> : IInterface<T>, IEquatable<IInterface<T>> {
    private readonly T[] _enumerable;

    public Class2(IEnumerable<T> enumerable) {
        var array = enumerable as T[] ?? enumerable.ToArray();
        _enumerable = array;
        Count = array.Length;
    }

    public IEnumerator<T> GetEnumerator() => ((IEnumerable<T>) _enumerable).GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public int Count { get; }
    public bool Equals(IInterface<T> other) => ReferenceEquals(this, other) || Count == other?.Count;

    public override bool Equals(object obj) {
        if (ReferenceEquals(this, obj)) return true;
        if (obj == null) return false;
        var that = obj as IInterface<T>;
        return that != null && Equals(that);
    }
}
```
Notice how `Class2` overrides `Equals()` and implements `IEquatable<IInterface<T>>`. When I create a test like the following, I would expect it to pass, since I only compare the `Count` property in this case:
```c#
[Test]
public void TestMethod() {
    var actual = new Class1<int>(new[] {1, 2, 3, 4, 5});
    var expected = new Class2<int>(new[] {5, 4, 3, 2, 1});
    Assert.That(actual, Is.EqualTo(expected));
}
```
The problem is the test fails, because the two classes are compared as `IEnumerable<T>`. This seems to contradict the documentation above. I would expect it to use `Class2`'s `Equals(IInterface<T>)` method first. Otherwise it should use its overridden `Equals(object)` method. Neither seems to be the case though.

**Comment by rprouse** [comment]
Does it work if you reverse the arguments to the assert? 

`Assert.That(expected, Is.EqualTo(actual));`

**Comment by lundmikkel** [comment]
No, it makes no difference. It still compares them as enumerables.

**Comment by lundmikkel** [comment]
And comparing it to itself also uses the enumerable:
```c#
Assert.That(expected, Is.EqualTo(expected));
```

**Comment by rprouse** [comment]
NUnitComparer has special code for enumerables and for iequatable. We do a number of tests and select the best option. I expect we are comparing using the enumerable code so are not getting to the IEquatable. I think IEquatable should take precedence.

**Comment by lundmikkel** [comment]
It seems that the problem is, as you said, that `NUnitEqualityComparer`'s `AreEqual` checks that the objects are enumerable before it checks for the proper `IEquatable` implementations. It seems that [lines 240-244](https://github.com/nunit/nunit/blob/43cf26b619ed08049f92c07318c29c80b523d28e/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L240-244) should be moved up to [line 155](https://github.com/nunit/nunit/blob/43cf26b619ed08049f92c07318c29c80b523d28e/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L155):

```C#
public bool AreEqual(object x, object y, ref Tolerance tolerance)
{
    this.failurePoints = new List<FailurePoint>();

    if (x == null && y == null)
        return true;

    if (x == null || y == null)
        return false;

    if (object.ReferenceEquals(x, y))
        return true;

    Type xType = x.GetType();
    Type yType = y.GetType();

    Type xGenericTypeDefinition = xType.GetTypeInfo().IsGenericType ? xType.GetGenericTypeDefinition() : null;
    Type yGenericTypeDefinition = yType.GetTypeInfo().IsGenericType ? yType.GetGenericTypeDefinition() : null;

    EqualityAdapter externalComparer = GetExternalComparer(x, y);
    if (externalComparer != null)
        return externalComparer.AreEqual(x, y);

    if (xType.IsArray && yType.IsArray && !compareAsCollection)
        return ArraysEqual((Array)x, (Array)y, ref tolerance);

    if (x is IDictionary && y is IDictionary)
        return DictionariesEqual((IDictionary)x, (IDictionary)y, ref tolerance);
    
    // Issue #70 - EquivalentTo isn't compatible with IgnoreCase for dictionaries
    if (x is DictionaryEntry && y is DictionaryEntry)
        return DictionaryEntriesEqual((DictionaryEntry)x, (DictionaryEntry)y, ref tolerance);

    // IDictionary<,> will eventually try to compare it's key value pairs when using CollectionTally
    if (xGenericTypeDefinition == typeof(KeyValuePair<,>) &&
        yGenericTypeDefinition == typeof(KeyValuePair<,>))
    {
        var keyTolerance = Tolerance.Exact;
        object xKey = xType.GetProperty("Key").GetValue(x, null);
        object yKey = yType.GetProperty("Key").GetValue(y, null);
        object xValue = xType.GetProperty("Value").GetValue(x, null);
        object yValue = yType.GetProperty("Value").GetValue(y, null);

        return AreEqual(xKey, yKey, ref keyTolerance) && AreEqual(xValue, yValue, ref tolerance);
    }

    //if (x is ICollection && y is ICollection)
    //    return CollectionsEqual((ICollection)x, (ICollection)y, ref tolerance);

    if (x is string && y is string)
        return StringsEqual((string)x, (string)y);

    if (x is IEnumerable && y is IEnumerable)
        return EnumerablesEqual((IEnumerable)x, (IEnumerable)y, ref tolerance);

    if (x is Stream && y is Stream)
        return StreamsEqual((Stream)x, (Stream)y);

    if ( x is char && y is char )
        return CharsEqual( (char)x, (char)y );

#if !PORTABLE
    if (x is DirectoryInfo && y is DirectoryInfo)
        return DirectoriesEqual((DirectoryInfo)x, (DirectoryInfo)y);
#endif

    if (Numerics.IsNumericType(x) && Numerics.IsNumericType(y))
        return Numerics.AreEqual(x, y, ref tolerance);

#if !NETCF
    if (x is DateTimeOffset && y is DateTimeOffset)
    {
        bool result;

        DateTimeOffset xAsOffset = (DateTimeOffset)x;
        DateTimeOffset yAsOffset = (DateTimeOffset)y;

        if (tolerance != null && tolerance.Value is TimeSpan)
        {
            TimeSpan amount = (TimeSpan)tolerance.Value;
            result = (xAsOffset - yAsOffset).Duration() <= amount;
        }
        else
        {
            result = xAsOffset == yAsOffset;
        }

        if (result && WithSameOffset)
        {
            result = xAsOffset.Offset == yAsOffset.Offset;
        }

        return result;
    }
#endif

    if (tolerance != null && tolerance.Value is TimeSpan)
    {
        TimeSpan amount = (TimeSpan)tolerance.Value;

        if (x is DateTime && y is DateTime)
            return ((DateTime)x - (DateTime)y).Duration() <= amount;

        if (x is TimeSpan && y is TimeSpan)
            return ((TimeSpan)x - (TimeSpan)y).Duration() <= amount;
    }

    MethodInfo equals = FirstImplementsIEquatableOfSecond(xType, yType);
    if (equals != null)
        return InvokeFirstIEquatableEqualsSecond(x, y, equals);
    if (xType != yType && (equals = FirstImplementsIEquatableOfSecond(yType, xType)) != null)
        return InvokeFirstIEquatableEqualsSecond(y, x, equals);
    
    return x.Equals(y);
}
```

**Comment by rprouse** [comment]
That is what I thought the fix would be. @nunit/core-team does anyone else see an issue with this fix? @lundmikkel can you submit a PR based on your fix and a couple of tests?

**Comment by CharliePoole** [comment]
Looks right to me. We should add tests using objects derived from each of the Types it now precedes.

This of course doesn't resolve the issue of simple Equals overrides that don't involve an interface. I think we have that as a separate issue.

**Comment by lundmikkel** [comment]
@rprouse I wouldn't mind doing it, but I must admit that the NUnit solution is so big, I have trouble figuring out which of the 18k tests to run. Several tests fail – and it seems like they should – so how do I know which ones to actually run and believe?

**Comment by CharliePoole** [comment]
@lundmikkel From your comment, I suspect you are using the VS Adapter or some other integrated runner that discovers all "tests" in the solution. That won't work when you are building NUnit because we have tests that are run at a second-level by other tests. We isolate them in separate assemblies to keep them out of the way, but most IDE-integrated solutions want to run them all. Best bet is to set up nunit3-console as your startup solution and use it to run the 4.5 framework build - since you are working on the framework. Then, before merging run tests from the command-line using build.

**Comment by lundmikkel** [comment]
@CharliePoole Yeah, exactly. Okay, let me have a look at that. I'll write again when everything fails :D

**Comment by lundmikkel** [comment]
Okay, so I think I've figured out which tests to run. One of the reason many of them failed, was because of the changed suggested above. It seems, it should be tackled in a different way: the enumerable check should be moved down instead. This leaves all tests still passing!

@rprouse @CharliePoole Do you have suggestions to where I should put the new tests? Can I use an existing class or should this rather be in a new spot? (I'll go to bed now, but I'll have a look at it again in the morning)

**Comment by CharliePoole** [comment]
@lundmikkel I'd just add them to the other NUnitEqualityComparerTests

**Comment by lundmikkel** [comment]
I've added some tests, that initially failed. I've tried to bundle them with similar tests and to match the coding style in the other tests. Let me know if I should change anything.

**Comment by CharliePoole** [comment]
Fixed by #1441 - please use proper form of issue references for auto-close.
